src/Microsoft.AspNet.Razor/Parser/HtmlMarkupParser.Block.cs

```diff
@@ -652,19 +688,43 @@ private void ParseQuoted(HtmlSymbolType type)
         private bool StartTag(Stack<Tuple<HtmlSymbol, SourceLocation>> tags, IDisposable tagBlockWrapper)
         {
             // If we're at text, it's the name, otherwise the name is ""
-            HtmlSymbol tagName;
-            if (At(HtmlSymbolType.Text))
+            HtmlSymbol tagName = null;
+            HtmlSymbol bangSymbol = null;
+            HtmlSymbol potentialTagNameSymbol;
+
+            if (At(HtmlSymbolType.Bang))
             {
-                tagName = CurrentSymbol;
+                bangSymbol = CurrentSymbol;
+
+                potentialTagNameSymbol = Lookahead();
             }
             else
             {
-                tagName = new HtmlSymbol(CurrentLocation, String.Empty, HtmlSymbolType.Unknown);
+                potentialTagNameSymbol = CurrentSymbol;
             }
 
+            if (potentialTagNameSymbol == null || potentialTagNameSymbol.Type != HtmlSymbolType.Text)
+            {
+                tagName = new HtmlSymbol(potentialTagNameSymbol.Start, string.Empty, HtmlSymbolType.Unknown);
+            }
+            else if (bangSymbol != null)
+            {
+                tagName = new HtmlSymbol(bangSymbol.Start, "!" + potentialTagNameSymbol.Content, HtmlSymbolType.Text);
+            }
+            else if (potentialTagNameSymbol.Type == HtmlSymbolType.Text)
+            {
+                tagName = potentialTagNameSymbol;
+            }
+
+            // tagName should never be null here, the above if/else if should take care of all cases.
+            Debug.Assert(tagName != null);
```

Comment:
don't repeat check of `potentialTagNameSymbol.Type`.  remove this `Assert()` and change last `else if` to an `else`